design-documents/checkout/alternative-checkout-flow.md

```diff
@@ -0,0 +1,160 @@
+     * Line Items:
+       * SKU
+       * Selected options (custom/configurable)
+       * Quantity
+       * Regular price
+       * Price
+  3. Quote factory arguments: Quote
+     * !Line Items
+     * Dimensions (weight & size) (based on LineItems)
+     * Shipping (requited for physical products)
+       * Address
+       * Selected Shipping Method
+     * Billing address (required for virtual products). Need use cases
+     * Coupons
+     * Gift cards
+     * Store credit
+     * Cart rules (calculated by Applicable catalog rules calculator)
+     * Customer
+  4. Totals calculator: Totals
+     * QuoteArgumentDTO
+     * PreviousTotals
+  5. Quote
+     * Line Items
+        * Dimensions
+     * Shipping address
+     * Price Adjustments
+        * Line Items Totals
+            * Totals
+            * Adjustments
+                * Taxes
+                * Cart rule discounts
+        * Coupons
+        * Gift cards
+        * Store credit
+        * Taxes
+     * Customer
+  6. Place Order: Order
+     * Quote
+     * PaymentMethod with Billing Addresses
+
+## Quote creation flow
+
+A quote should be created from Cart after a customer clicks `Proceed to Checkout`, on the API level it will be a separate service contract like `QuoteEstimator::create`. After the quote is created, it became an immutable object and only totals can be recalculated (the totals will be represented as a separate object). If a list of items should be changed, when a new quote object should be created. Quote will live until TTL expires or an order is successfully placed.
+
+![Quote Calculation](img/alternative-quote-calculation.png)
+
+## Totals calculation improvements
+
+The current approach for quote calculation has multiple drawbacks like changes in the quote object, a quote totals collector is difficult to customize, a complicated logic to define the order of totals calculation (tax before/after discount, discount/tax rules for shipping, etc.), additional calls to 3rd party systems for shipping rate prices updating.
+
+The proposed solution assumes that a quote will be an immutable object and will operate with totals list, each calculator will create add new totals object based on totals list and the order of each calculation can be changed in runtime. As one of the benefits of the proposed approached - the list of calculators and their order can be visualized for better calculation understanding.
+
+Let's consider the calculation of the totals might look like.
+
+![Totals Calculation](img/totals-calculation-pipeline.png)
+
+Each calculator receives Quote DTO and Totals List, calculates totals, creates new Totals DTO with the calculated amount and adds it to Totals List. This approach allows to do not change a quote object, have a defined interface for totals and change the order of calculation. Magento provides multiple configurations to change the order of calculation, for example, discount can be applied before shipping amount or after.
+
```

Comment:
One common customization for the current totals calculation is that custom discounts are implemented by adjusting prior totals, e.g. the subtotal, from a later total model. If the totals are immutable, there should be an option for total models to return a changed list of totals, e.g. remove or replace total models that where calculated by prior. Each total model then can be immutable, but can be replaced or removed as needed by customizations.

Reply:
Not sure if I clearly understand your point but totals calculators will be created in runtime according to the configuration and priority and each calculator has `isApplicable` method (the customization can disable needed calculator, add a new one) which decides if calculator should be executed or not, for example, it does not make sense to execute a shipping calculator for the quote only with virtual products.

A custom discount calculator can be added in the same way as standard calculators with own priority and grand total calc will automatically collect all amounts and adjustments.